jerry-core/api/jerry.c

```diff
@@ -465,16 +470,55 @@ jerry_parse (const jerry_char_t *resource_name_p, /**< resource name (usually a
 
   if (JERRY_UNLIKELY (bytecode_data_p == NULL))
   {
+#if ENABLED (JERRY_MODULE_SYSTEM)
+    if ((parse_opts & JERRY_PARSE_MODULE) != 0)
+    {
+      ecma_module_cleanup_context ();
+    }
+#endif /* ENABLED (JERRY_MODULE_SYSTEM) */
+
     return ecma_create_error_reference_from_context ();
   }
 
+#if ENABLED (JERRY_MODULE_SYSTEM)
+  if ((parse_opts & JERRY_PARSE_MODULE) != 0)
+  {
+    ecma_module_t *root_module_p = JERRY_CONTEXT (module_current_p);
+
+    if (ECMA_IS_VALUE_ERROR (ecma_module_parse_referenced_modules ()))
+    {
+      ecma_bytecode_deref (bytecode_data_p);
+      ecma_module_cleanup_context ();
+
+      return ecma_create_error_reference_from_context ();
+    }
+
+    ecma_object_t *obj_p = ecma_create_object (NULL, sizeof (ecma_extended_object_t), ECMA_OBJECT_TYPE_CLASS);
```

Comment:
What happens if you execute this with postfix `()` in js?

Reply:
You would get a TypeError, these objects are not callable. The only way to execute these is to pass them to jerry_run.

Reply:
Can they be executed twice? If not, would it be better to execute them here?
When the referenced modules are executed?

Reply:
The main module can be executed multiple times. The referenced modules are executed only once before the main module is first evaluated, and are cached for subsequent runs.
This could be changed so that the whole module tree is evaluated each time the main module is executed, but in that case we would need to keep the bytecode for the referenced modules, which we currently don't do to conserve memory.

Reply:
Looking at it a bit, subsequent runs of the main module will (almost) always result in an error, since the imported variables will already exist, so it doesn't make a lot of sense to be able to run them multiple times. 

However, I would like to keep parsing and execution separate.